common/src/chain/config.rs

```diff
@@ -47,6 +55,9 @@ pub struct ChainConfig {
     genesis_block_id: Id<Block>,
     blockreward_maturity: BlockDistance,
     version: SemVer,
+    target_block_spacing: Duration,
+    coin_decimals: u8,
+    emission_schedule: Vec<(BlockHeight, Amount)>,
```

Comment:
Hmm we need total emission as well, maybe it could be a function that gives the total supply at given block height and the reward would be calculated as total supply at current block minus total supply at previous block.

Reply:
Unfortunately, while the function idea is programmatically a good idea and I do like it, it doesn't fulfill the requirement for a generic mapping that we can convert from a string. In this paradigm, the requirement is to be able to parse the schedule from a string. This helps in creating a functional test that tests the schedule without having to mine years worth of blocks.

Reply:
About the function with the total supply; we can't do that here. The correct way to do this (which will come up later) is whenever we accept a block in the chain, there will be a total supply entry in the block index that tells us the total supply at that block. This is, unfortunately, not achievable abstractly here because miners/stakers are free to renounce their rewards. Mining a block with zero reward is totally acceptable.

Reply:
To avoid off-by-one errors, I added lots of tooling and used pairs instead. Check it out.